toolchain/parse/typed_nodes.h

```diff
@@ -0,0 +1,781 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#ifndef CARBON_TOOLCHAIN_PARSE_TYPED_NODES_H_
+#define CARBON_TOOLCHAIN_PARSE_TYPED_NODES_H_
+
+#include "toolchain/parse/node_ids.h"
+#include "toolchain/parse/node_kind.h"
+
+namespace Carbon::Parse {
+
+// A pair of a list item and its optional following comma.
+template <typename Element, typename Comma>
+struct ListItem {
+  Element value;
+  std::optional<Comma> comma;
+};
+
+// A list of items, parameterized by the kind of the elements and comma.
+template <typename Element, typename Comma>
+using CommaSeparatedList = llvm::SmallVector<ListItem<Element, Comma>>;
+
+// This class provides a shorthand for defining parse node kinds for leaf nodes.
+template <const NodeKind& KindT, NodeCategory Category = NodeCategory::None>
+struct LeafNode {
+  static constexpr auto Kind = KindT.Define(Category);
+};
+
+// Each type defined below corresponds to a parse node kind, and describes the
+// expected child structure of that parse node.
+
+// Error nodes
```

Comment:
I think it would be helpful to do something to make these subheadings stand out a bit more in the file. Given that we use markdown-inspired syntax in our comments already, how about one of these:
```suggestion
// ### Error nodes
```
or
```suggestion
// Error nodes
// -----------
```

Reply:
Used the latter